[WIP] added elasticity to composite viscoplastic

bobmyhill opened the pull request:
This PR adds elasticity to the composite viscoplastic material model. The corresponding schematic is shown below:
![image](https://user-images.githubusercontent.com/8144905/125843843-b87853cd-8677-44e0-96a8-926f804ebee0.png)

* [ ] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).
* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

naliboff commented on include/aspect/material_model/rheology/composite_visco_plastic.h:
```diff
@@ -108,14 +111,28 @@ namespace aspect
+                                                    const PeierlsCreepParameters peierls_creep_parameters,
+                                                    const DruckerPragerParameters drucker_prager_parameters) const;
+
+          /**
```
I'm a bit confused about the need for this function as `elasticity->strain_rates_from_stored_elastic_stresses` is the main operation done within it along with a check to make sure elasticity is used. Can you provide a brief explanation above about why a direct call the function in the elasticity module would not work?

bobmyhill replied:
I think you're right. It might not even be needed at all.

naliboff reviewed (commented):
@bobmyhill - Exciting, almost at the full VEP composite rheology! 

I had a quick look through and left a series of general questions and comments. I want to go through the reordering in the elastic module in detail (for my understanding), but otherwise everything is in very good shape as always. 

More comments soon, thank you again!

naliboff commented:
@bobmyhill - Also, we will need to eventually document carefully (in multiple places) that this is a fundamental departure from the VEP method/ordering currently used in the visco plastic model. In fact, the ordering of elements is in line with the original implementation we used (elasticity after VP).

bobmyhill commented on source/material_model/rheology/composite_visco_plastic.cc:
```diff
@@ -227,13 +240,12 @@ namespace aspect
 
         // Compute the strain rate experienced by the different mechanisms
-        // These should sum to the total strain rate
+        // These should sum to the total strain rate iff elasticity is not used.
```
Hah! In this case I really meant *iff* (as in *if and only if*). But I'll change it to the longer form.

bobmyhill commented on include/aspect/material_model/rheology/composite_visco_plastic.h:
```diff
@@ -108,14 +111,28 @@ namespace aspect
-                                              const DislocationCreepParameters dislocation_creep_parameters,
-                                              const PeierlsCreepParameters peierls_creep_parameters,
-                                              const DruckerPragerParameters drucker_prager_parameters) const;
+          compute_creep_strain_rate_and_derivative (const double creep_stress,
```
Yes, that's the reason - this function deals only with the four "real" viscoplastic mechanisms. *creep* was the shortest useful name I could think of.

bobmyhill commented on include/aspect/material_model/rheology/composite_visco_plastic.h:
```diff
@@ -80,7 +81,8 @@ namespace aspect
                              const std::vector<double> &phase_function_values = std::vector<double>(),
-                             const std::vector<unsigned int> &n_phases_per_composition = std::vector<unsigned int>()) const;
+                             const std::vector<unsigned int> &n_phases_per_composition = std::vector<unsigned int>(),
+                             const double shear_modulus = 0.) const;
```
Good catch, I'll change this.

bobmyhill commented:
@naliboff thank you for the early review! 
I think that maybe we don't need to restructure the elasticity module in this PR. It will need to be changed though; at the moment the code segfaults when I try to run with elasticity turned on. The problem is on the line `const unsigned int n_fields = this->n_compositional_fields() + 1;` in Elasticity::parse_parameters.

naliboff commented on source/material_model/rheology/composite_visco_plastic.cc:
```diff
@@ -183,22 +187,31 @@ namespace aspect
+            double strain_rate = creep_stress/(2.*max_viscosity) + (max_viscosity/(max_viscosity - min_viscosity))*creep_edot_and_deriv.first;
             strain_rate_deriv = 1./(2.*max_viscosity) + (max_viscosity/(max_viscosity - min_viscosity))*creep_edot_and_deriv.second;
 
+            if (use_elasticity)
```
Great that this all that was needed (for now) to incorporate elasticity! Certainly easier when elasticity is out of the "creep" circuit.

naliboff reviewed (commented):
@bobmyhill - Thanks for addressing the comments so quickly! Should I hold off on further review until a test is added?

bobmyhill commented:
> @bobmyhill - Thanks for addressing the comments so quickly! Should I hold off on further review until a test is added?

Yes, this isn't going to work until we have sorted out the elasticity module.

bobmyhill commented:
Will close this and open new PR.
